test/framework/exec/docker.go

```diff
@@ -0,0 +1,360 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	httpProxy  = "HTTP_PROXY"
+	httpsProxy = "HTTPS_PROXY"
+	noProxy    = "NO_PROXY"
+)
+
+var dockerClient *client.Client
+
+// getDockerClient returns a new client connection for interacting with the Docker engine.
+func getDockerClient() (*client.Client, error) {
+	if dockerClient == nil {
+		var err error
+		dockerClient, err = client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+		if err != nil {
+			return nil, err
```

Comment:
maybe it would be good to add some error context here. Some of the error I see in there aren't very self-explanatory, e.g. `return nil, fmt.Errorf("unable to verify TLS configuration, invalid transport %v", c.client.Transport)`

Reply:
It's hard to give too explicit of a message there without checking for all the different ways that could fail. I've added a little more context around the error now, so maybe with that context it will be a little easier to understand what's going on if something were to fail here.